### ISSUE TITLE
mtp_generate_step: logits processors see stale prev_tokens on draft calls

### ISSUE THREAD

**Issue by Thump604** [ml-explore/mlx-lm#1175]
## Problem

`mtp_generate_step` in `generate.py` calls `_process_and_sample(prev_tokens, mtp_logits)` at line 769 (`_step_mtp`), but `prev_tokens` does not contain the backbone's just-sampled token. A stateful logits processor that reads `tokens[-1]` to determine state (e.g. Outlines' Guide FSM for constrained JSON decoding) advances by the wrong token and enters an invalid state.

Serial `generate_step` does not have this issue because the sampled token is fed back as `input_tokens` on the next `_step` call, so `tokens[-1]` always reflects the last generated token.

## Reproduction

Script below runs a logits processor through both `generate_step` and `mtp_generate_step`, logging what `tokens[-1]` is on each call:

```python
import mlx.core as mx
from mlx_lm import load
from mlx_lm.generate import generate_step, mtp_generate_step

MODEL = "path/to/mtp-model"  # any model with mtp_forward

def make_tracker(tokenizer):
    tok = tokenizer._tokenizer
    log = []
    n = [0]
    def processor(tokens, logits):
        n[0] += 1
        tlist = tokens.tolist() if hasattr(tokens, "tolist") else list(tokens)
        if isinstance(tlist, int): tlist = [tlist]
        log.append((n[0], tlist[-1], len(tlist), tok.decode([tlist[-1]])))
        return logits
    return processor, log

model, tokenizer = load(MODEL)
tok = tokenizer._tokenizer

# --- Serial ---
proc, log = make_tracker(tokenizer)
gen = list(zip(range(6), generate_step(model=model,
    prompt=mx.array(tokenizer.encode("Hello world")),
    max_tokens=6, logits_processors=[proc])))
print("Serial tokens[-1] per call:", [(c, t, d) for c, t, _, d in log])

# --- MTP ---
proc2, log2 = make_tracker(tokenizer)
gen2 = list(zip(range(6), mtp_generate_step(model=model,
    prompt=mx.array(tokenizer.encode("Hello world")),
    max_tokens=6, logits_processors=[proc2])))
print("MTP tokens[-1] per call:", [(c, t, d) for c, t, _, d in log2])
```

**Output on Qwen3.5-27B-VLM-MTP-8bit:**

Serial (7 calls, 6 tokens -- correct):
```
call 1: tokens[-1]=1814(' world'), len=1  (prompt, first call)
call 2: tokens[-1]=0('!'),        len=2  (generated token from call 1)
call 3: tokens[-1]=271('\n\n'),    len=3  (generated token from call 2)
call 4: tokens[-1]=9419('Hello'),  len=4
call 5: tokens[-1]=0('!'),        len=5
call 6: tokens[-1]=2500(' How'),   len=6
call 7: tokens[-1]=628(' can'),    len=7
```

MTP (10 calls, 6 tokens -- draft calls see stale tokens):
```
call 1:  tokens[-1]=1814(' world'), len=2  (prompt, first call)
call 2:  tokens[-1]=1814(' world'), len=2  ← MTP draft: STALE (should be '!' from call 1)
call 3:  tokens[-1]=0('!'),         len=3
call 4:  tokens[-1]=0('!'),         len=4  (n_predict=2 verify)
call 5:  tokens[-1]=0('!'),         len=4  ← MTP draft: STALE
call 6:  tokens[-1]=9419('Hello'),   len=5
call 7:  tokens[-1]=9419('Hello'),   len=6
call 8:  tokens[-1]=9419('Hello'),   len=6  ← MTP draft: STALE
call 9:  tokens[-1]=2500(' How'),    len=7
call 10: tokens[-1]=2500(' How'),    len=8
```

Calls 2, 5, 8 are MTP draft calls where `prev_tokens` has not been updated with the backbone's sampled token.

## Root cause

In `_step_backbone` (line 747-760), `prev_tokens` grows by `y_ctx` (the backbone input tokens), and the sampled token is returned but NOT appended to `prev_tokens`. When `_step_mtp` (line 762-770) then calls `_process_and_sample(prev_tokens, mtp_logits)`, the processor sees `prev_tokens[-1]` as the backbone's input, not its output.

For serial `generate_step`, this is not a problem because `_step` is called once per token, and the sampled token becomes the next call's `input_tokens` which gets concatenated into `tokens` at line 422-425.

## Impact

Blocks all stateful logits processors from working with MTP generation. Outlines' Guide FSM, which advances state via `tokens[-1]` on each call (see `outlines.backends.outlines_core`, line 188: `self._guides[i].advance(token_id=last_token_id)`), enters an invalid FSM state and produces "No next state found" errors.

Stateless processors (repetition penalty, temperature scaling) are unaffected.

## Environment

- mlx-lm 0.31.2
- mlx 0.27+
- outlines 1.2.12 / outlines_core 0.2.14
- macOS, Apple Silicon M2 Ultra

**Comment by lawcontinue** [comment]
Hit this building MTP speculative decoding on a sharded MLX pipeline (Gemma3-12B across two Mac Minis, Thunderbolt link).

The fix: append backbone token t to prev_tokens *before* calling the MTP head. We measured MTP overhead at 9.2ms/draft token (vs 101ms full forward), so the draft distribution has to match the verifier exactly — otherwise you burn 9ms on a rejected token.

The FSM desync makes sense: draft sees stale state, proposes a token valid for position n-1 but illegal at n, and the FSM has no recovery. Same root cause as acceptance rate tanking — wrong conditional distribution because it's missing the backbone's sampled token.

Clean reproduction. The fix should be one line in `_step_mtp` to extend prev_tokens before `_process_and_sample`.

**Comment by Thump604** [comment]
The fixes in #990 (commit 67c10e6) address all three bugs reported here:

1. Fixed y_ctx slice: per-iteration `y[i:i+1]` instead of reusing `y[:-(n_predict-1)]`
2. Local `tokens_for_proc` with `main_tok` appended, avoiding `prev_tokens` mutation
3. `prev_tokens = prev_tokens[:-1]` after rejection to trim the rejected draft token

Agreed that `nonlocal prev_tokens` would double-count since `_step_backbone` already appends at `i=0`. The local copy approach is cleaner.

The stateful processor rollback limitation is real and affects our thinking budget work at #1196. Will open a dedicated issue for that -- it needs a `save_state()`/`restore_state()` protocol on the processor interface to be correct with any speculative path.

This issue can be closed once #990 merges.

**Comment by Thump604** [comment]
I have a fix for this in our vendored mlx-lm 0.31.2. Two changes in `generate.py`:

**1. Per-position `y[i:i+1]` in the verification loop**

The verification loop used the same `y_ctx` for all positions, which meant `confirmed_tok` was doubled in `prev_tokens` and `draft_tok` was missing:

```diff
-            y_ctx = y if n_predict == 1 else y[: -(n_predict - 1)]
             for i in range(n_predict):
+                y_i = y[i : i + 1] if n_predict > 1 else y
                 prev_tokens = (
-                    mx.concatenate([prev_tokens, y_ctx])
+                    mx.concatenate([prev_tokens, y_i])
                 )
```

**2. Snapshot/restore `prev_tokens` on draft rejection**

On rejection, `prev_tokens` was never rolled back — it accumulated the draft token's processor state permanently:

```diff
+                _prev_tokens_before_verify = prev_tokens
                 toks, lps, hidden = _step_backbone(
                     y_with_draft, n_predict=2, n_confirmed=1
                 )
                 ...
                 else:
                     # Draft rejected
+                    if logits_processors:
+                        prev_tokens = mx.concatenate(
+                            [_prev_tokens_before_verify, y]
+                        ) if _prev_tokens_before_verify is not None else y
```

This applies to the 0.31.2 MTP path (`_step_backbone`/`_step_mtp`). I see the current main has restructured to a different draft model architecture. If the current code still processes `prev_tokens` across speculative positions, the same bug class applies — `prev_tokens` must track the actual confirmed token at each position, not a shared context.

Evidence: with this fix, essay repetition collapse dropped from 40% to 10% on Qwen 3.6 35B-A3B.
